### ISSUE TITLE
关于Fig.17的可视化

### ISSUE THREAD

**Issue by renqi1998** [mengduanjinghua/SS-MTr#1]
您好，
        十分感谢您有意义的工作，我在阅读您的论文时，发现你在Fig.17可视化出注意力图，觉得十分新颖，您可以分享该部分可视化代码吗？

**Comment by mengduanjinghua** [comment]
![image](https://github.com/mengduanjinghua/SS-MTr/assets/45927629/e587d113-4323-4f16-96dd-4c79da906726)
我画的注意力图是根据上面的过程进行的。
具体来说，首先需要Transformer里attention计算模块返回attention值，也就是q*k。
class Attention(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0.):
        super().__init__()
        self.num_heads = num_heads
        head_dim = dim // num_heads
        self.scale = qk_scale or head_dim ** -0.5

        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x):
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
        q, k, v = qkv[0], qkv[1], qkv[2]
        attn = (q @ k.transpose(-2, -1)) * self.scale  # B x num_heads x N x N
        attn = attn.softmax(dim=-1)

        attn = self.attn_drop(attn)
        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
        x = self.proj(x)
        x = self.proj_drop(x)
        return x, attn

然后在模型前向传播的时候，将attention保存：
attn_weights = []
for i, blk in enumerate(self.blocks):
    x, weights = blk(x)
    if len(self. Blocks) - i <= n:
        attn_weights.append(weights)


 attn_weights = torch. stack(attn_weights) #维度是  层数 * batchsize *head个数 * (HW+1) * (HW+1)

最后画图：
def plot_head_attntion(img_rgb, attn):
    weights = np.transpose(attn, (1,2,0,3,4))  #bs * num_head * layer * (HW+1) * (HW+1)
    patch_weights = weights[:,:,:,1:,1:]  #bs * num_head * layer * HW * HW
    b, head, l, hw, _ = patch_weights.shape
    h = int(np.sqrt(hw))
    w = int(np.sqrt(hw))
    cls_weights = weights[:,:,:,0,1:]   #bs * num_head * layer * HW

    cls_weights = cls_weights.reshape(b, head, l, h, w)  #bs * num_head * layer * H * W
    cam_ = cls_weights[:,:,-1,:,:]

    for i in range(b):
        cam = cam_[i]
        cam = cam - np.min(cam)
        cam_head = cam / (np.max(cam)+1e-10)
        rgb = img_rgb[i] 
        rgb = np.uint8(255*(rgb - np.min(rgb))/(np.max(rgb)-np.min(rgb))+1e-10)
        for j in range(head):
            cam_img = cam_head[j,:,:]
            # cam = cam - np.min(cam)
            # cam_img = cam / (np.max(cam)+1e-10)
            cam_img = np.uint8(255 * cam_img)
            heatmap = cv2.applyColorMap(cv2.resize(cam_img,(27, 27)), cv2.COLORMAP_JET)
            # result = heatmap * 0.3 + rgb * 0.5     #将热力图叠加到原图上
            result = heatmap
            cv2.imwrite(‘路径’, result）#这个画出的是每个样本每个head得到的注意力图
        return 0

**Comment by mengduanjinghua** [comment]
> 您好， 十分感谢您有意义的工作，我在阅读您的论文时，发现你在Fig.17可视化出注意力图，觉得十分新颖，您可以分享该部分可视化代码吗？

![image](https://github.com/mengduanjinghua/SS-MTr/assets/45927629/e587d113-4323-4f16-96dd-4c79da906726)
我画的注意力图是根据上面的过程进行的。
具体来说，首先需要Transformer里attention计算模块返回attention值，也就是q*k。

```
class Attention(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0.):
        super().__init__()
        self.num_heads = num_heads
        head_dim = dim // num_heads
        self.scale = qk_scale or head_dim ** -0.5
        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x):
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
        q, k, v = qkv[0], qkv[1], qkv[2]
        attn = (q @ k.transpose(-2, -1)) * self.scale  # B x num_heads x N x N
        attn = attn.softmax(dim=-1)

        attn = self.attn_drop(attn)
        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
        x = self.proj(x)
        x = self.proj_drop(x)
        return x, attn
```

然后在模型前向传播的时候，将attention保存：
```
attn_weights = []
for i, blk in enumerate(self.blocks):
    x, weights = blk(x)
    if len(self. Blocks) - i <= n:
        attn_weights.append(weights)

 attn_weights = torch. stack(attn_weights) #维度是  层数 * batchsize *head个数 * (HW+1) * (HW+1)
```

最后画图：
```
def plot_head_attntion(img_rgb, attn):
    weights = np.transpose(attn, (1,2,0,3,4))  #bs * num_head * layer * (HW+1) * (HW+1)
    patch_weights = weights[:,:,:,1:,1:]  #bs * num_head * layer * HW * HW
    b, head, l, hw, _ = patch_weights.shape
    h = int(np.sqrt(hw))
    w = int(np.sqrt(hw))
    cls_weights = weights[:,:,:,0,1:]   #bs * num_head * layer * HW

    cls_weights = cls_weights.reshape(b, head, l, h, w)  #bs * num_head * layer * H * W
    cam_ = cls_weights[:,:,-1,:,:]

    for i in range(b):
        cam = cam_[i]
        cam = cam - np.min(cam)
        cam_head = cam / (np.max(cam)+1e-10)
        rgb = img_rgb[i] 
        rgb = np.uint8(255*(rgb - np.min(rgb))/(np.max(rgb)-np.min(rgb))+1e-10)
        for j in range(head):
            cam_img = cam_head[j,:,:]
            # cam = cam - np.min(cam)
            # cam_img = cam / (np.max(cam)+1e-10)
            cam_img = np.uint8(255 * cam_img)
            heatmap = cv2.applyColorMap(cv2.resize(cam_img,(27, 27)), cv2.COLORMAP_JET)
            # result = heatmap * 0.3 + rgb * 0.5     #将热力图叠加到原图上
            result = heatmap
            cv2.imwrite(‘路径’, result）#这个画出的是每个样本每个head得到的注意力图
        return 0

```
